src/pyfaf/retrace.py

```diff
@@ -98,62 +99,88 @@ def __init__(self, db_debug_package, db_src_package, bin_pkg_map, db=None):
 # pylint: enable-msg=R0903
 
 
-class RetraceWorker(threading.Thread, object):
+class RetracePool:
     """
-    The worker providing asynchronous unpacking of packages.
+    A class representing a pool of workers that run the retracing job for given tasks.
     """
 
-    def __init__(self, worker_id, inqueue, outqueue):
-        name = "Worker #{0}".format(worker_id)
-        super(RetraceWorker, self).__init__(name=name)
-        self.inqueue = inqueue
-        self.outqueue = outqueue
-        self.stop = False
-        # Instance of 'RootLogger' has no 'getChildLogger' member
+    def __init__(self, db, tasks, problemplugin, workers):
+
+        self.name = "RetracePool"
         # pylint: disable-msg=E1103
-        self.log = log.getChildLogger("{0}.{1}".format(self.__class__.__name__,
-                                                       self.name))
+        self.log = log.getChildLogger(self.name)
         # pylint: enable-msg=E1103
 
-    def _process_task(self, task):
+        self.db = db
+        self.plugin = problemplugin
+        self.tasks = tasks
+        self.total = len(tasks)
+        self.workers = workers
+
+    def run(self):
+        """
+        Starts the executors job and schedules tasks for workers when they are available.
         """
-        Asynchronously unpack one set of packages (debuginfo, source, binary)
+        taskid = 0
+
+        with futures.ThreadPoolExecutor(max_workers=self.workers, thread_name_prefix=self.name) as executor:
+            while self.tasks:
+                taskid += 1
+                task = self.tasks.popleft()
+                try:
+                    future = executor.submit(self._process_task, task, taskid, self.db)
```

Comment:
`_process_task()` does not take a database parameter.

Reply:
:man_facepalming: